spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubInboundChannelAdapter.java

```diff
@@ -145,24 +145,26 @@ public void nack() {
 			});
 		}
 
-		boolean messageNacked = false;
-
 		try {
 			sendMessage(getMessageBuilderFactory()
 					.withPayload(message.getPayload())
 					.copyHeaders(messageHeaders)
 					.build());
+
+			if (this.ackMode == AckMode.AUTO_ACK || this.ackMode == AckMode.AUTO) {
+				message.ack();
+			}
 		}
 		catch (RuntimeException re) {
 			if (this.ackMode == AckMode.AUTO) {
 				message.nack();
-				messageNacked = true;
 			}
-			throw new PubSubException("Sending Spring message failed.", re);
-		}
-		finally {
-			if (!messageNacked && (this.ackMode == AckMode.AUTO || this.ackMode == AckMode.AUTO_ACK)) {
-				message.ack();
+
+			if (this.ackMode == AckMode.AUTO_ACK) {
+				LOGGER.warn("Sending Spring message failed.", re);
```

Comment:
AUTO_ACK behavior needs to be the same as MANUAL when an exception is thrown -- rethrow, so the user can handle it.

Reply:
+1, why would we not want to rethrow the exception for `AUTO_ACK`?

Reply:
I went off the request made by @radivojen in the original issue that we should consider not throwing here -

> throwing PubSubException causes immediate redelivery by the google pubsub client

I'm not really sure what the Spring cloud stream specs call for here.

Reply:
How is throwing the exception causing immediate redelivery? Pub/Sub will generally redeliver only after the  ack-deadline expires. However, it's possible for a duplicate message to be delivered sooner because Pub/Sub only guarantees at-least-once delivery.

Reply:
If the exception is propagated it has to be handled by the Google pubsub sdk.
Since the subscribe message flow is googlesdk -> spring-gcp
Here is the log I receive, handled by google sdk as WARN level:

> Radivojes-MacBook-Pro.local 2019-12-12 16:17:12,427 reqId= userAgent= WARN  com.google.cloud.pubsub.v1.MessageDispatcher ->  [gcp-pubsub-subscriber1] - MessageReceiver failed to processes ack ID: K0MLKlgRTgQhIT4wPkVTRFAGFixdRkhRNxkIaFEOT14jPzUgKEUSIysdFmNdNBsNaFxcdAEEW0gjK2cnbgkZAQhFe059ZBwNalRcdAJVCB53dWh2aHLOlrPUiv31agExUJTfjNhkLdDQ1cJEZh09WBJLLD5-LSxFQV5AEkwmGURJUw, the message will be nacked.

Reply:
And here is the exact method from the MessageDispatcher:
`@Override
    public void onFailure(Throwable t) {
      logger.log(
          Level.WARNING,
          "MessageReceiver failed to processes ack ID: " + ackId + ", the message will be nacked.",
          t);
      pendingNacks.add(ackId);
      forget();
    }`

On the publisher level it is of course fine to propagate the exception.
I hope I helped. Please test it yourselves.

Reply:
You're right. Essentially, it ends up behaving just like the `AUTO` mode, which is not the intent.